### DISCUSSION TITLE
How to parallelise tests

### DISCUSSION THREAD

**Discussion by mehulkar** [prysk/prysk#166] [Q&A] [answered]
It would be nice to run multiple tests at the same time to make them faster. I'm not sure, but it seems `TESTDIR` is unique for tests, so there is some isolation built in already that would prevent tests from running into each other.

**Comment by Nicoretti** [comment]
Hi @mehulkar,

I think this is a solved issue see https://github.com/Nicoretti/prysk/discussions/128#discussioncomment-4950209.

best
Nico

**Comment by Nicoretti** [comment, marked as answer]
@mehulkar prysk tests can be executed in parallel, by using the `prysk-pytest` plugin together with `pytest` and the `pytest-xdist` plugin (`prysk >= 0.13.0`). 

Usage:

1. Install `prysk` with `pytest-plugin`
```shell
pip install "prysk[pytest-plugin]"
``` 

2. Install `pytest-xdist` for parallel test execution 
```shell
pip install pytest-xdist
```

3. Execute your `prysk` tests using `pytest` in parallel
```shell
pytest -n auto /Path/To/Your/Test/Folder
```

**Reply by mehulkar** [reply]
This is great. Is it possible to pass the `-i` flag using pytest also?

**Reply by Nicoretti** [reply]
@mehulkar currently only a small subset of the options is available when using the plugin see [here](https://github.com/Nicoretti/prysk/blob/0ad7550e83a20523d0bcda670c97e7dce0269093/pytest_prysk/__init__.py#L50) or run `pytest --help | grep prysk`. Generally we can be extend the options, the `-i` flag is a special one though imho (especially when doing parallel execution).

Can you give me some more details on the **how** and **why** you would want to use it when running the tests with `pytest`?

**Comment by mehulkar** [comment]
One thing I noticed while migrating tests here is that the names of tmp directories no longer match the prysk test, so some tests that assert against those paths fail. I don't think this is a big deal, but wanted to flag for others that may be running into it.

**Reply by Nicoretti** [reply]
@mehulkar this is a good point 👍. I haven't been aware of that.

**_thanks!_**

That said let me add a few notes from my perspective:

1. Generally I would not advise to rely on (deducing) specific names for the temporary folder.
If the name of the temporary folder is needed, I think users should use the value of `$PRYSK_TEMP`.

2. Still, if the temp folder behaviour of the pytest execution imposes a non trivial problem to anyone, **please let me know!** 
Making the behaviour compliant with "vanilla" prysk shouldn't be to complicated. 

3. If nobody has an issue with that change, I would use the opportunity to "drop this feature" (not support it via pytest).


P.S.: @mehulkar thanks once more for the upstream contributions (be it feedback and/or PR(s)), I really appreciate it!

**Reply by mehulkar** [reply]
`$PRYSK_TEMP` could help, but I don't have a way to expand the variable in the asserted log output.

**Reply by Nicoretti** [reply]
> `$PRYSK_TEMP` could help, but I don't have a way to expand the variable in the asserted log output.

@mehulkar [this PR](https://github.com/Nicoretti/prysk/pull/177) focuses on expanding `$TMPDIR` in the output, but this maybe already is sufficient for your use case too.
